### PR TITLE
[OV20] Preprocessing: User friendly error message for convert_layout

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
 - Fix incomprehensible error message during layout conversion when layout rank doesn't match with shape rank
 - Example exception error message: "Conversion layout [N,C,?,?,?] <-> [N,?,?,?,C] failure. Layout is not consistent with input shape {?,3,2,1}. Layout length 5 shall match with input shape rank 4"

### Tickets:
 - 71867

### CODE REVIEW

**Comment by mgrdv** [comment]
Now it looks good for me

File: src/core/tests/preprocess.cpp [hunk 1]
```diff
@@ -534,6 +534,31 @@ TEST(pre_post_process, reuse_model_layout_no_tensor_info) {
     EXPECT_EQ(f->get_parameters().front()->get_layout(), "NC??");
 }
 
+TEST(pre_post_process, set_layout_out_of_bounds) {
+    auto shape = PartialShape{Dimension::dynamic(), 3, 2, 1};
+    std::stringstream shape_str;
+    shape_str << shape;
+    auto f = create_simple_function(element::f32, shape);
+    Layout from{"N???C"};
+    Layout to{"NC???"};
```
Comment (ilyachur) [inline src/core/tests/preprocess.cpp:543]: Could you add next test:
```
PartialShape p{1, 2};
Layout {"NCHW"}
```

Comment (nosovmik) [reply src/core/tests/preprocess.cpp]: Done

File: src/core/include/openvino/core/layout.hpp [hunk 2]
```diff
@@ -88,7 +90,7 @@ class OPENVINO_API Layout {
     friend Layout layout::apply_permutation(const Layout& src_layout, const std::vector<uint64_t>& dims);
 
     friend std::vector<int64_t> layout::find_permutation(const Layout& src_layout,
-                                                         const Rank& src_shape_rank,
+                                                         const PartialShape& src_shape,
                                                          const Layout& dst_layout);
 };
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/core/layout.hpp:95]: BTW, can we move these functions to `src`? They are not exported, but still usable by users of static OpenVINO libraries

**Review by ilya-lavrenov — APPROVED** [review]

File: src/core/include/openvino/core/layout.hpp (hunk 2 above)
Comment (nosovmik) [reply src/core/include/openvino/core/layout.hpp:80]: Done. But to hide function signatures from 'friend declaration' I had to move implementation into 'class LayoutUtils'

**Review by ilyachur — APPROVED** [review]

File: src/core/tests/preprocess.cpp [hunk 3]
```diff
@@ -534,6 +534,31 @@ TEST(pre_post_process, reuse_model_layout_no_tensor_info) {
...
 }
 
+TEST(pre_post_process, set_layout_out_of_bounds) {
+    auto shape = PartialShape{1, 2};
+    std::stringstream shape_str;
+    shape_str << shape;
+    auto f = create_simple_function(element::f32, shape);
+    Layout from{"NHWC"};
+    Layout to{"NCHW"};
+    // TODO: replace with EXPECT_THAT after upgrade gtest to v1.11
```
Comment (ilya-lavrenov) [inline src/core/tests/preprocess.cpp:544]: BTW, you can try to migrate to v1.11 using this state https://github.com/openvinotoolkit/googletest/tree/release-1.11.0-openvino

**Pull request merged**